server/linkpreview.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"golang.org/x/net/html"
+	"golang.org/x/net/html/atom"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+	"unicode/utf8"
+)
+
+type linkPreview struct {
+	Title       string `json:"title,omitempty"`
+	Description string `json:"description,omitempty"`
+	ImageURL    string `json:"image_url,omitempty"`
+}
+
+var client = &http.Client{
+	Timeout: time.Second * 2,
+	CheckRedirect: func(req *http.Request, via []*http.Request) error {
+		if err := validateURL(req.URL); err != nil {
+			return err
+		}
+		return nil
+	},
+}
+
+// previewLink handles the HTTP request, fetches the URL, and returns the link preview.
+func previewLink(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet && r.Method != http.MethodHead {
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// check authorization
+	uid, challenge, err := authHttpRequest(r)
+	if err != nil {
+		http.Error(w, "invalid auth secret", http.StatusBadRequest)
+		return
+	}
+	if challenge != nil || uid.IsZero() {
+		http.Error(w, "user not authenticated", http.StatusUnauthorized)
+		return
+	}
+
+	u := r.URL.Query().Get("url")
+	if u == "" {
+		http.Error(w, "Missing 'url' query parameter", http.StatusBadRequest)
+		return
+	}
+
+	pu, err := url.Parse(u)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	if err := validateURL(pu); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	req, err := http.NewRequest(http.MethodGet, u, nil)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadGateway)
+		return
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices { // StatusCode != 20X
+		http.Error(w, "Non-OK HTTP status", http.StatusBadGateway)
+		return
+	}
+
+	body := http.MaxBytesReader(nil, resp.Body, 2*1024) // 2KB limit
+	if cc := resp.Header.Get("Cache-Control"); cc != "" {
+		w.Header().Set("Cache-Control", cc)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if r.Method == http.MethodHead {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
```

Comment:
Completely missed it earlier: please check `resp.Header.Get("Content-Type") == "text/html"`. If it's not text/html, then there is no need to parse the response. Support for `image/*` and `video/*` can be added later.

Reply:
like this?
```go
if strings.Contains(resp.Header.Get("Content-Type"), "text/html") {
		if err := json.NewEncoder(w).Encode(extractMetadata(body)); err != nil {
			http.Error(w, "Failed to encode response", http.StatusInternalServerError)
		}
	}
}
```

Example: `Content-Type: text/html; charset=utf-8`

Reply:
Should not we move the trimspace at the end, for len of char less than the limit, trimspace will not applied.
```go
if utf8.RuneCountInString(preview.Title) > 80 {
		preview.Title = string([]rune(preview.Title)[:80])
	}
	if utf8.RuneCountInString(preview.Description) > 256 {
		preview.Description = string([]rune(preview.Description)[:256])
	}
	if len(preview.ImageURL) > 2000 {
		preview.ImageURL = preview.ImageURL[:2000]
	}
	
	return &linkPreview{
		Title:       strings.TrimSpace(preview.Title),
		Description: strings.TrimSpace(preview.Description),
		ImageURL:    strings.TrimSpace(preview.ImageURL),
	}
```